src/catalog/dependency_manager.cpp

```diff
@@ -27,166 +115,311 @@ void DependencyManager::AddObject(CatalogTransaction transaction, CatalogEntry &
+}
+
+static bool CascadeDrop(bool cascade, DependencyType dependency_type) {
+	if (cascade) {
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_AUTOMATIC) {
+		// These dependencies are automatically dropped implicitly
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_OWNS) {
+		// The object has explicit ownership over the dependency
+		return true;
+	}
+	return false;
+}
+
+void DependencyManager::UnmangleName(const string &mangled, CatalogType &type, string &schema, string &name) {
+	auto parts = StringUtil::Split(mangled, std::string("\0", 1));
+	D_ASSERT(parts.size() == 3);
+	type = CatalogTypeFromString(parts[0]);
+	schema = std::move(parts[1]);
+	name = std::move(parts[2]);
+}
+
+void GetLookupProperties(CatalogEntry &entry, string &schema, string &name, CatalogType &type) {
+	if (entry.type == CatalogType::DEPENDENCY_ENTRY) {
+		auto &dependency_entry = entry.Cast<DependencyCatalogEntry>();
+
+		schema = dependency_entry.EntrySchema();
+		name = dependency_entry.EntryName();
+		type = dependency_entry.EntryType();
+	} else if (entry.type == CatalogType::DEPENDENCY_SET) {
+		auto &dependency_set = entry.Cast<DependencySetCatalogEntry>();
+
+		auto &mangled_name = dependency_set.MangledName();
+		DependencyManager::UnmangleName(mangled_name, type, schema, name);
+	} else {
+		throw InternalException("Unrecognized CatalogType in 'GetLookupProperties'");
+	}
+}
+
+// Always performs the callback, it's up to the callback to determine what to do based on the lookup result
+optional_ptr<CatalogEntry> DependencyManager::LookupEntry(CatalogTransaction transaction, CatalogEntry &dependency) {
+	string schema;
+	string name;
+	CatalogType type;
+	GetLookupProperties(dependency, schema, name, type);
+
+	// Lookup the schema
+	auto schema_entry = catalog.schemas->GetEntry(transaction, schema);
+	if (type == CatalogType::SCHEMA_ENTRY || !schema_entry) {
+		// This is a schema entry, perform the callback only providing the schema
+		return schema_entry;
+	}
+	auto &duck_schema_entry = schema_entry->Cast<DuckSchemaEntry>();
+
+	// Lookup the catalog set
+	auto &catalog_set = duck_schema_entry.GetCatalogSet(type);
+
+	// Use the index to find the actual entry
+	auto entry = catalog_set.GetEntry(transaction, name);
+	return entry;
+}
+
+void DependencyManager::CleanupDependencies(CatalogTransaction transaction, CatalogEntry &object) {
+	auto connections_p = GetDependencySet(transaction, object);
+	D_ASSERT(connections_p);
+	auto &connections = *connections_p;
+
+	// Collect the dependencies
+	catalog_entry_set_t dependencies_to_remove;
+	connections.ScanDependencies(transaction, [&](DependencyCatalogEntry &dep) { dependencies_to_remove.insert(dep); });
+	// Also collect the dependents
+	catalog_entry_set_t dependents_to_remove;
+	connections.ScanDependents(transaction, [&](DependencyCatalogEntry &dep) { dependents_to_remove.insert(dep); });
+
+	// Remove the dependency entries
```

Comment:
Maybe we should also take this opportunity to clean up the unconnected DEPENDENCY_OWNS link between object -> entry